### PR TITLE
add grounded-sam

### PR DESCRIPTION
**Pull request by LokeZhou**
* add sam
* add sam deploy
* add grounded-sam

### CODE REVIEW

File: applications/inpaint/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+## Grounded-SAM: Detect and Segment Everything with Text Prompt
+
+## Prepare
+
+```bash
+# install
+pip install paddlenlp
```
Comment (jerrywgz) [inline applications/inpaint/README.md]: 不需要额外安装paddlenlp了

File: applications/CVinW/grounded-sam.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
+from dataclasses import dataclass, field
```
Comment (jerrywgz) [inline applications/CVinW/grounded-sam.py]: python文件名规范 grounded_sam.py
这里需要兼容高性能推理能力，或者新增脚本

File: applications/CVinW/grounded-sam.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        },
+    )
+
+def main():
+    parser = PdArgumentParser((ModelArguments, DataArguments))
+    model_args, data_args = parser.parse_args_into_dataclasses()
+
+    #bulid dino processor
+    dino_processor = GroudingDinoProcessor.from_pretrained(
+        'bert-base-uncased'
```
Comment (jerrywgz) [inline applications/CVinW/grounded-sam.py]: 这里的processor是否可以和模型名绑定，bert和groundingdino本身并不直接相关

File: applications/CVinW/grounded-sam.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    ) 
+
+    #bulid dino model
+    logger.info("dino_model: {}".format(model_args.dino_model_name_or_path))
+    dino_model = GroundingDinoModel.from_pretrained(model_args.dino_model_name_or_path)
+    
+    #buidl sam processor
+    sam_processor = SamProcessor.from_pretrained(
+        'Sam'
```
Comment (jerrywgz) [inline applications/CVinW/grounded-sam.py]: 同上，和model_args.sam_model_name_or_path 绑定？

File: applications/CVinW/grounded-sam.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    ) 
+    #bulid model
+    logger.info("SamModel: {}".format(model_args.sam_model_name_or_path))
+    sam_model = SamModel.from_pretrained(model_args.sam_model_name_or_path,input_type="boxs")
+
+    #read image
+    image_pil = Image.open(data_args.input_image).convert("RGB")
```
Comment (jerrywgz) [inline applications/CVinW/grounded-sam.py]: 可以参考blip2的run_predict.py 支持url

File: applications/CVinW/grounded-sam.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    seg_masks = sam_model(img=image_seg,prompt=prompt)
+    seg_masks = sam_processor.postprocess_masks(seg_masks)
+    
+    if model_args.visual:
+        # make dir
+        os.makedirs(model_args.output_dir, exist_ok=True)
+        init_mask = postprocess(seg_masks)
+        
+        image_masked = mask_image(image_pil, init_mask)
+        image_masked.save(os.path.join(model_args.output_dir, "image_masked.jpg"))
```
Comment (jerrywgz) [inline applications/CVinW/grounded-sam.py]: add logger here

File: deploy/groundingdino/predict.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+        self.input_map['image'] = self.image.numpy()
+        self.input_map['mask'] = np.array(self.mask.numpy(),dtype='int64')
+
+        for key in self.tokenized_input.keys():
+            self.input_map[key] = np.array(self.tokenized_input[key].numpy(),dtype='int64')
+       
+        input_names = self.predictor.get_input_names()
+        for i in range(len(input_names)):
+            input_tensor = self.predictor.get_input_handle(input_names[i])
+            if input_names[i] == 'x':
```
Comment (jerrywgz) [inline deploy/groundingdino/predict.py]: 这里可以统一导出key和input-map的key 简化下代码

File: deploy/sam/predict.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+        data_args.points_prompt = np.array([data_args.points_prompt])
+    
+    if use_auto_tune(model_args):
+        tune_img_nums = 10
+        auto_tune(model_args, [image_pil], tune_img_nums)
+
+    predictor = Predictor(model_args)
+    
+    image_pil = Image.open(data_args.input_image).convert("RGB")
+    seg_masks = predictor.run(image_pil,
```
Comment (jerrywgz) [inline deploy/sam/predict.py:335]: sam和grounded sam的pipeline要统一下，建议钱后处理都封装到predictor.run里

File: paddlevlp/models/sam/mask_decoder.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        b, c, h, w = src.shape
+        hs, src = self.transformer(src, pos_src, tokens)
+
+        iou_token_out = hs[:, (0), :]
+        mask_tokens_out = hs[:, 1:1 + self.num_mask_tokens, :]
+        x = src
+        perm_0 = list(range(x.ndim))
+        perm_0[1] = 2
+        perm_0[2] = 1
+        """Class Method: *.view, not convert, please check whether it is torch.Tensor.*/Optimizer.*/nn.Module.*, and convert manually"""
```
Comment (jerrywgz) [inline paddlevlp/models/sam/mask_decoder.py]: 去除这类注释

File: applications/CVinW/grounded-sam.py (hunk 2 above)
Comment (LokeZhou) [reply applications/CVinW/grounded-sam.py]: 动态图兼容，静态图待优化完成后加脚本

File: applications/CVinW/grounded-sam.py (hunk 3 above)
Comment (LokeZhou) [reply applications/CVinW/grounded-sam.py]: done

File: applications/CVinW/grounded-sam.py (hunk 4 above)
Comment (LokeZhou) [reply applications/CVinW/grounded-sam.py]: done

File: applications/CVinW/grounded-sam.py (hunk 5 above)
Comment (LokeZhou) [reply applications/CVinW/grounded-sam.py]: done

File: applications/CVinW/grounded-sam.py (hunk 6 above)
Comment (LokeZhou) [reply applications/CVinW/grounded-sam.py]: done

File: deploy/groundingdino/predict.py (hunk 7 above)
Comment (LokeZhou) [reply deploy/groundingdino/predict.py]: done

File: deploy/sam/predict.py (hunk 8 above)
Comment (LokeZhou) [reply deploy/sam/predict.py:335]: done

File: paddlevlp/models/sam/mask_decoder.py (hunk 9 above)
Comment (LokeZhou) [reply paddlevlp/models/sam/mask_decoder.py]: done

**Review by jerrywgz — APPROVED** [review]
LGTM

File: applications/CVinW/grounded-sam.py (hunk 5 above)
Comment (LokeZhou) [reply applications/CVinW/grounded-sam.py]: done
